File: impl/php/src/Directory.php
<?php
declare(strict_types=1);

namespace Nyx;

/**
 * Verzeichnis: Eintraege pruefen, Kette fuehren.
 *
 * Die Pruefregeln sind dieselben wie in nyx/directory.py. Wichtig ist vor
 * allem die zweite: die Adresse ist der Schluessel. Ein Eintrag, dessen
 * Adresse sich nicht aus seinem Identitaetsschluessel ergibt, ist wertlos,
 * egal wie gueltig seine Signatur ist.
 */
final class Directory
{
    private const RECORD_LABEL = 'nyx/v1/record';
    private const BLOCK_LABEL = 'nyx/v1/block';
    private const LEAF = "\x00";
    private const NODE = "\x01";
    private const TYPES = ['BIND', 'PREKEY', 'REVOKE', 'RECOVER'];

    private string $path;

    public function __construct(string $dataDir = '/tmp/nyx-store', private int $bits = 12)
    {
        @mkdir($dataDir, 0700, true);
        $this->path = rtrim($dataDir, '/') . '/chain.json';
        if (!is_file($this->path)) {
            $this->save([$this->mine([
                'height' => 0, 'prev' => str_repeat('0', 64),
                'merkle_root' => str_repeat('0', 64), 'ts' => 0,
                'bits' => $this->bits, 'nonce' => 0, 'records' => [],
            ])]);
        }
    }

    // -- Adressen ---------------------------------------------------------

    public static function addressFromKey(string $ikPub): string
    {
        $body = substr(Canonical::sha256('nyx/v1/addr', $ikPub), 0, 20);
        $sum = substr(Canonical::sha256('nyx/v1/addrsum', $body), 0, 2);
        return strtolower(rtrim(self::base32($body . $sum), '='));
    }

    private static function base32(string $raw): string
    {
        $alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        $bits = '';
        for ($i = 0, $n = strlen($raw); $i < $n; $i++) {
            $bits .= str_pad(decbin(ord($raw[$i])), 8, '0', STR_PAD_LEFT);
        }
        $out = '';
        foreach (str_split($bits, 5) as $chunk) {
            $out .= $alphabet[bindec(str_pad($chunk, 5, '0'))];
        }
        return $out;
    }

    // -- Eintraege ---------------------------------------------------------

    public static function verifyRecord(array $record): bool
    {
        if (!in_array($record['type'] ?? '', self::TYPES, true)) {
            return false;
        }
        foreach (['addr', 'ik_pub', 'sig', 'ts', 'body'] as $field) {
            if (!array_key_exists($field, $record)) {
                return false;
            }
        }
        $ikPub = @hex2bin((string) $record['ik_pub']);
        $sig = @hex2bin((string) $record['sig']);
        if ($ikPub === false || $sig === false || strlen($ikPub) !== 32 || strlen($sig) !== 64) {
            return false;
        }
        if (self::addressFromKey($ikPub) !== $record['addr']) {
            return false;
        }

        $unsigned = $record;
        unset($unsigned['sig'], $unsigned['recovery_sig'], $unsigned['height']);
        $message = self::RECORD_LABEL . Canonical::encode($unsigned);
        return sodium_crypto_sign_verify_detached($sig, $message, $ikPub);
    }

    // -- Kette -------------------------------------------------------------

    public function blocks(): array
    {
        return json_decode((string) file_get_contents($this->path), true) ?: [];
    }

    private function save(array $blocks): void
    {
        file_put_contents($this->path, json_encode($blocks), LOCK_EX);
    }

    public function headers(): array
    {
        return array_map(fn(array $b) => self::header($b), $this->blocks());
    }

    private static function header(array $block): array
    {
        return [
            'height' => $block['height'], 'prev' => $block['prev'],
            'merkle_root' => $block['merkle_root'], 'ts' => $block['ts'],
            'bits' => $block['bits'], 'nonce' => $block['nonce'],
        ];
    }

    public static function blockHash(array $block): string
    {
        return bin2hex(Canonical::sha256(self::BLOCK_LABEL,
            Canonical::encode(self::header($block))));
    }

    public static function merkleRoot(array $records): string
    {
        if (!$records) {
            return str_repeat("\0", 32);
        }
        $level = array_map(
            fn(array $r) => Canonical::sha256(self::LEAF, Canonical::encode($r)),
            $records
        );
        while (count($level) > 1) {
            if (count($level) % 2) {
                $level[] = end($level);
            }
            $next = [];
            for ($i = 0; $i < count($level); $i += 2) {
                $next[] = Canonical::sha256(self::NODE, $level[$i], $level[$i + 1]);
            }
            $level = $next;
        }
        return $level[0];
    }

    private function mine(array $block): array
    {
        while (Canonical::leadingZeroBits(
            Canonical::sha256(self::BLOCK_LABEL, Canonical::encode(self::header($block)))
        ) < $block['bits']) {
            $block['nonce']++;
        }
        return $block;
    }

    public function submit(array $record, bool $seal = true): int
    {
        if (!self::verifyRecord($record)) {
            throw new \InvalidArgumentException('Eintrag ist nicht gueltig signiert');
        }
        $blocks = $this->blocks();
        $prev = $blocks[count($blocks) - 1];

        $block = $this->mine([
            'height' => count($blocks),
            'prev' => self::blockHash($prev),
            'merkle_root' => bin2hex(self::merkleRoot([$record])),
            'ts' => time(),
            'bits' => $this->bits,
            'nonce' => 0,
            'records' => [$record],
        ]);
        $blocks[] = $block;
        $this->save($blocks);
        return $block['height'];
    }

    public function validate(): bool
    {
        $blocks = $this->blocks();
        foreach ($blocks as $i => $block) {
            if ($block['height'] !== $i) {
                return false;
            }
            if ($i > 0 && $block['prev'] !== self::blockHash($blocks[$i - 1])) {
                return false;
            }
            if (Canonical::leadingZeroBits(
                Canonical::sha256(self::BLOCK_LABEL, Canonical::encode(self::header($block)))
            ) < $block['bits']) {
                return false;
            }
            if ($block['merkle_root'] !== bin2hex(self::merkleRoot($block['records']))) {
                return false;
            }
            foreach ($block['records'] as $record) {
                if (!self::verifyRecord($record)) {
                    return false;
                }
            }
        }
        return true;
    }

    // -- Lesen -------------------------------------------------------------

    public function recordsFor(string $address, ?string $type = null): array
    {
        $out = [];
        foreach ($this->blocks() as $block) {
            foreach ($block['records'] as $record) {
                if ($record['addr'] === $address && ($type === null || $record['type'] === $type)) {
                    $out[] = $record + ['height' => $block['height']];
                }
            }
        }
        return $out;
    }

    private function revoked(string $address): array
    {
        $keys = [];
        foreach ($this->recordsFor($address, 'REVOKE') as $record) {
            if (isset($record['body']['key'])) {
                $keys[] = $record['body']['key'];
            }
        }
        return $keys;
    }

    public function resolve(string $address): ?array
    {
        $revoked = $this->revoked($address);
        foreach (array_reverse($this->recordsFor($address, 'BIND')) as $record) {
            $idk = $record['body']['idk_pub'] ?? null;
            if ($idk !== null && !in_array($idk, $revoked, true)) {
                return ['address' => $address, 'ik_pub' => $record['ik_pub'], 'idk_pub' => $idk];
            }
        }
        return null;
    }

    public function latestBundle(string $address): ?array
    {
        $revoked = $this->revoked($address);
        foreach (array_reverse($this->recordsFor($address, 'PREKEY')) as $record) {
            if (!in_array($record['body']['spk_pub'] ?? '', $revoked, true)) {
                return $record['body'];
            }
        }
        return null;
    }

    /**
     * Jeder Schluesselwechsel mit Blockhoehe und Zeitpunkt.
     *
     * Der praktische Nutzen der Kette: ein untergeschobener Schluessel
     * erscheint hier als zusaetzlicher Eintrag und ist damit sichtbar,
     * ohne dass jemand Pruefsummen von Hand vergleichen muesste.
     */
    public function keyHistory(string $address): array
    {
        return array_map(fn(array $r) => [
            'height' => $r['height'], 'type' => $r['type'], 'ts' => $r['ts'],
            'key' => $r['body']['idk_pub'] ?? $r['body']['spk_pub'] ?? $r['body']['key'] ?? null,
        ], $this->recordsFor($address));
    }

    public function height(): int
    {
        return count($this->blocks()) - 1;
    }
}

File: impl/python/nyx/directory.py
"""Das Verzeichnis: eine nur anhaengbare Kette von Schluesselereignissen.

In die Kette gehen keine Nachrichten. Nur vier Eintragsarten:

  BIND     bindet einen X25519-Handshakeschluessel an eine Identitaet
  PREKEY   veroeffentlicht ein signiertes Prekey-Buendel
  REVOKE   widerruft einen Schluessel
  RECOVER  benennt eine Nachfolge-Identitaet

Der Zweck ist nicht, das Unterschieben eines Schluessels unmoeglich zu
machen, sondern es sichtbar und dauerhaft nachweisbar zu machen. Wer einen
fremden Schluessel unterschiebt, hinterlaesst einen Eintrag, den jeder
sehen und niemand nachtraeglich entfernen kann — genau das leistet
``key_history()``.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass, field

from .identity import Identity, PublicIdentity, address_matches_key, verify_signature
from .primitives import sha256

BLOCK_LABEL = b"nyx/v1/block"
RECORD_LABEL = b"nyx/v1/record"
LEAF_LABEL = b"\x00"
NODE_LABEL = b"\x01"

RECORD_TYPES = ("BIND", "PREKEY", "REVOKE", "RECOVER")


def canonical(obj) -> bytes:
    """Eindeutige Byte-Darstellung, damit Python, PHP und JS gleich hashen."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"),
                      ensure_ascii=True).encode("ascii")


# ------------------------------------------------------------- Eintraege

def make_record(identity: Identity, rtype: str, body: dict, ts: int | None = None) -> dict:
    if rtype not in RECORD_TYPES:
        raise ValueError(f"unbekannte Eintragsart: {rtype}")
    record = {
        "type": rtype,
        "addr": identity.address,
        "ik_pub": identity.ik_pub.hex(),
        "ts": int(ts if ts is not None else time.time()),
        "body": body,
    }
    record["sig"] = identity.sign(RECORD_LABEL + canonical(unsigned_part(record))).hex()
    return record


def verify_record(record: dict) -> bool:
    """Signatur, Adressbindung und Aufbau — alles oder nichts."""
    try:
        if record.get("type") not in RECORD_TYPES:
            return False
        ik_pub = bytes.fromhex(record["ik_pub"])
        sig = bytes.fromhex(record["sig"])
        if not address_matches_key(record["addr"], ik_pub):
            # Die Adresse ist der Schluessel. Passt das nicht, ist der
            # Eintrag wertlos, egal wie gueltig die Signatur ist.
            return False
        return verify_signature(ik_pub, RECORD_LABEL + canonical(unsigned_part(record)), sig)
    except (KeyError, ValueError, TypeError):
        return False


def unsigned_part(record: dict) -> dict:
    """Der signierte Umfang eines Eintrags: alles ausser den Signaturen."""
    return {k: v for k, v in record.items() if k not in ("sig", "recovery_sig")}


def verify_recover(record: dict, recovery_pub: bytes) -> bool:
    """Eine Nachfolge braucht beide Unterschriften: die der Identitaet und
    die des getrennt aufbewahrten Wiederherstellungsschluessels."""
    if record.get("type") != "RECOVER" or not verify_record(record):
        return False
    try:
        sig = bytes.fromhex(record["recovery_sig"])
    except (KeyError, ValueError):
        return False
    return verify_signature(recovery_pub, RECORD_LABEL + canonical(unsigned_part(record)), sig)


def record_hash(record: dict) -> bytes:
    return sha256(LEAF_LABEL, canonical(record))


# ----------------------------------------------------------- Merkle-Baum

def merkle_root(records: list[dict]) -> bytes:
    if not records:
        return bytes(32)
    level = [record_hash(r) for r in records]
    while len(level) > 1:
        if len(level) % 2:
            level.append(level[-1])
        level = [sha256(NODE_LABEL, level[i], level[i + 1])
                 for i in range(0, len(level), 2)]
    return level[0]


def merkle_proof(records: list[dict], index: int) -> list[dict]:
    """Pfad, mit dem ein Leichtclient einen Eintrag gegen den Blockkopf
    prueft, ohne alle Eintraege zu laden."""
    level = [record_hash(r) for r in records]
    proof: list[dict] = []
    while len(level) > 1:
        if len(level) % 2:
            level.append(level[-1])
        sibling = index ^ 1
        proof.append({"side": "left" if sibling < index else "right",
                      "hash": level[sibling].hex()})
        level = [sha256(NODE_LABEL, level[i], level[i + 1])
                 for i in range(0, len(level), 2)]
        index //= 2
    return proof


def verify_merkle_proof(record: dict, proof: list[dict], root: bytes) -> bool:
    node = record_hash(record)
    for step in proof:
        sibling = bytes.fromhex(step["hash"])
        node = (sha256(NODE_LABEL, sibling, node) if step["side"] == "left"
                else sha256(NODE_LABEL, node, sibling))
    return node == root


# ---------------------------------------------------------------- Bloecke

def block_hash(header: dict) -> bytes:
    return sha256(BLOCK_LABEL, canonical(header))


def leading_zero_bits(digest: bytes) -> int:
    n = 0
    for byte in digest:
        if byte:
            return n + (8 - byte.bit_length())
        n += 8
    return n


@dataclass
class Block:
    height: int
    prev: str
    merkle_root: str
    ts: int
    bits: int
    nonce: int = 0
    records: list[dict] = field(default_factory=list)

    def header(self) -> dict:
        return {
            "height": self.height,
            "prev": self.prev,
            "merkle_root": self.merkle_root,
            "ts": self.ts,
            "bits": self.bits,
            "nonce": self.nonce,
        }

    def hash(self) -> str:
        return block_hash(self.header()).hex()

    def to_dict(self) -> dict:
        d = self.header()
        d["records"] = self.records
        return d

    @staticmethod
    def from_dict(d: dict) -> "Block":
        return Block(d["height"], d["prev"], d["merkle_root"], d["ts"],
                     d["bits"], d["nonce"], d.get("records", []))


class Chain:
    """Kette mit Arbeitsnachweis.

    Die Schwierigkeit ist hier klein gewaehlt, damit die Demo auf einem
    Rechner laeuft. Im Betrieb ergibt sie sich aus der Rechenleistung des
    Netzes; die Pruefregeln bleiben dieselben.
    """

    def __init__(self, bits: int = 12):
        self.bits = bits
        self.blocks: list[Block] = []
        self.pending: list[dict] = []
        self._genesis()

    def _genesis(self) -> None:
        g = Block(0, "0" * 64, bytes(32).hex(), 0, self.bits, 0, [])
        self._mine(g)
        self.blocks.append(g)

    def _mine(self, block: Block) -> None:
        while leading_zero_bits(block_hash(block.header())) < block.bits:
            block.nonce += 1

    # -- Schreiben -------------------------------------------------------

    def submit(self, record: dict) -> None:
        """Ein Eintrag kostet Arbeit; ungueltige werden nicht angenommen."""
        if not verify_record(record):
            raise ValueError("Eintrag ist nicht gueltig signiert")
        self.pending.append(record)

    def seal(self, ts: int | None = None) -> Block | None:
        """Schliesst die anstehenden Eintraege in einen Block ein."""
        if not self.pending:
            return None
        records, self.pending = self.pending, []
        block = Block(
            height=len(self.blocks),
            prev=self.blocks[-1].hash(),
            merkle_root=merkle_root(records).hex(),
            ts=int(ts if ts is not None else time.time()),
            bits=self.bits,
            records=records,
        )
        self._mine(block)
        self.blocks.append(block)
        return block

    # -- Pruefen ---------------------------------------------------------

    def validate(self) -> bool:
        """Vollstaendige Pruefung, wie sie ein neuer Knoten durchfuehrt."""
        for i, block in enumerate(self.blocks):
            if block.height != i:
                return False
            if i and block.prev != self.blocks[i - 1].hash():
                return False
            if leading_zero_bits(block_hash(block.header())) < block.bits:
                return False
            if block.merkle_root != merkle_root(block.records).hex():
                return False
            if any(not verify_record(r) for r in block.records):
                return False
        return True

    def headers(self) -> list[dict]:
        """Was ein Leichtclient laedt — Kopf ohne Eintraege."""
        return [b.header() for b in self.blocks]

    def proof_for(self, address: str, rtype: str) -> dict | None:
        """SPV-Beweis fuer den juengsten Eintrag einer Art."""
        for block in reversed(self.blocks):
            for i, rec in enumerate(block.records):
                if rec["addr"] == address and rec["type"] == rtype:
                    return {
                        "record": rec,
                        "proof": merkle_proof(block.records, i),
                        "header": block.header(),
                    }
        return None

    # -- Lesen -----------------------------------------------------------

    def records_for(self, address: str, rtype: str | None = None) -> list[dict]:
        out = []
        for block in self.blocks:
            for rec in block.records:
                if rec["addr"] == address and (rtype is None or rec["type"] == rtype):
                    out.append({**rec, "height": block.height})
        return out

    def revoked_keys(self, address: str) -> set[str]:
        return {r["body"]["key"] for r in self.records_for(address, "REVOKE")
                if "key" in r.get("body", {})}

    def resolve(self, address: str) -> PublicIdentity | None:
        """Aktuelle Handshake-Bindung einer Adresse."""
        revoked = self.revoked_keys(address)
        for rec in reversed(self.records_for(address, "BIND")):
            idk = rec["body"]["idk_pub"]
            if idk not in revoked:
                return PublicIdentity(address, bytes.fromhex(rec["ik_pub"]),
                                      bytes.fromhex(idk))
        return None

    def latest_bundle(self, address: str) -> dict | None:
        revoked = self.revoked_keys(address)
        for rec in reversed(self.records_for(address, "PREKEY")):
            if rec["body"]["spk_pub"] not in revoked:
                return rec["body"]
        return None

    def key_history(self, address: str) -> list[dict]:
        """Jeder Schluesselwechsel, mit Blockhoehe und Zeitpunkt.

        Das ist der praktische Nutzen der Kette: Ein untergeschobener
        Schluessel erscheint hier als zusaetzlicher Eintrag. Ein Client, der
        diese Liste anzeigt, macht den Angriff fuer den Nutzer sichtbar,
        ohne dass dieser Pruefsummen von Hand vergleichen muss.
        """
        out = []
        for rec in self.records_for(address):
            body = rec.get("body", {})
            key = body.get("idk_pub") or body.get("spk_pub") or body.get("key")
            out.append({
                "height": rec["height"],
                "type": rec["type"],
                "ts": rec["ts"],
                "key": key,
            })
        return out


# --------------------------------------------------- Eintraege erzeugen

def bind_record(identity: Identity) -> dict:
    return make_record(identity, "BIND", {
        "idk_pub": identity.idk_pub.hex(),
        "recovery_pub": identity.recovery_pub.hex() if identity.recovery_pub else None,
    })


def prekey_record(identity: Identity, bundle_body: dict) -> dict:
    return make_record(identity, "PREKEY", bundle_body)


def revoke_record(identity: Identity, key_hex: str, reason: str = "") -> dict:
    return make_record(identity, "REVOKE", {"key": key_hex, "reason": reason})


def recover_record(identity: Identity, successor_address: str) -> dict:
    rec = make_record(identity, "RECOVER", {"successor": successor_address})
    rec["recovery_sig"] = identity.sign_recovery(
        RECORD_LABEL + canonical(unsigned_part(rec))).hex()
    return rec
